pkg/controller/mattermostcluster/mattermostcluster_controller.go

```diff
@@ -100,54 +110,524 @@ func (r *ReconcileMattermostCluster) Reconcile(request reconcile.Request) (recon
-	if err != nil && errors.IsNotFound(err) {
-		reqLogger.Info("Creating a new Pod", "Pod.Namespace", pod.Namespace, "Pod.Name", pod.Name)
-		err = r.client.Create(context.TODO(), pod)
+	if mattermost.Spec.DatabaseType.Type == "mysql" {
+		reqLogger.Info("Reconciling MattermostCluster Database MySQL service account")
+		err = r.checkDBServiceAccount(mattermost, reqLogger)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		reqLogger.Info("Reconciling MattermostCluster Database MySQL role binding")
+		err = r.checkDBRoleBinding(mattermost, reqLogger)
 		if err != nil {
 			return reconcile.Result{}, err
 		}
 
-		// Pod created successfully - don't requeue
-		return reconcile.Result{}, nil
-	} else if err != nil {
+		reqLogger.Info("Reconciling MattermostCluster Database MySQL")
+		err = r.checkDBMySQLDeployment(mattermost, reqLogger)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	} else {
+		reqLogger.Info("Reconciling MattermostCluster Database Postgres")
+		err = r.checkDBPostgresDeployment(mattermost, reqLogger)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+
+	// TODO
+	err = r.checkMinioDeployment(mattermost, reqLogger)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	err = r.checkMattermostDeployment(mattermost, reqLogger)
+	if err != nil {
 		return reconcile.Result{}, err
 	}
 
-	// Pod already exists - don't requeue
-	reqLogger.Info("Skip reconcile: Pod already exists", "Pod.Namespace", found.Namespace, "Pod.Name", found.Name)
+	// Already exists - don't requeue
 	return reconcile.Result{}, nil
 }
 
-// newPodForCR returns a busybox pod with the same name/namespace as the cr
-func newPodForCR(cr *mattermostv1alpha1.MattermostCluster) *corev1.Pod {
-	labels := map[string]string{
-		"app": cr.Name,
+func (r *ReconcileMattermostCluster) setDefaults(mattermost *mattermostv1alpha1.MattermostCluster, reqLogger logr.Logger) error {
+	reqLogger.Info("Checking defaults")
+	reqLogger.Info(mattermost.Spec.IngressName)
+	if mattermost.Spec.IngressName == "" {
+		return fmt.Errorf("need to set the IngressName")
+	}
+
+	if len(mattermost.Spec.Image) == 0 {
+		reqLogger.Info("Setting default Mattermost image: " + constants.DefaultMattermostImage)
+		mattermost.Spec.Image = constants.DefaultMattermostImage
+	}
+
+	if mattermost.Spec.Replicas == 0 {
+		reqLogger.Info("Setting default Mattermost replicas: " + strconv.Itoa(constants.DefaultAmountOfPods))
+		mattermost.Spec.Replicas = constants.DefaultAmountOfPods
+	}
+
+	if len(mattermost.Spec.DatabaseType.Type) == 0 {
+		reqLogger.Info("Setting default Mattermost database type: " + constants.DefaultMattermostDatabaseType)
+		mattermost.Spec.DatabaseType.Type = constants.DefaultMattermostDatabaseType
+	}
+
+	return nil
+}
+
+func (r *ReconcileMattermostCluster) checkDBServiceAccount(mattermost *mattermostv1alpha1.MattermostCluster, reqLogger logr.Logger) error {
```

Comment:
To me this method does three things:
1. Creates/gets the service account spec
2. Creates the service account, if necessary
3. Updates the service account, if necessary (not implemented yet)

My guess is that we (or someone else) may want to do these at different times so I'd suggest creating three smaller functions that `checkDBServiceAccount` can call.

1. `func (c *MattermostCluster) GetServiceAccountSpec() *corev1.ServiceAccount` - this one can move into the `v1alpha1` package as it fits there better and can be used outside reconciling
2. `func (r *ReconcileMattermostCluster) createServiceAccount(mattermost *mattermostv1alpha1.MattermostCluster, reqLogger logr.Logger) error`
3. `func (r *ReconcileMattermostCluster) updateServiceAccount(mattermost *mattermostv1alpha1.MattermostCluster, reqLogger logr.Logger) error`

Then `checkDBServiceAccount` can call `createServiceAccount`/`updateServiceAccount` as needed. Both of those could also call `GetServiceAccountSpec` themselves to get the service account spec or we could have that passed in, 0/5 on that

Reply:
remove the updateServiceAccount, because we dont have anything to update, at least for now